feat(link): add update endpoint for link

TheoPascoli opened the pull request:
Add an endpoint to update link

skamril committed:
v2.16.8

Merge pull request #2014 from AntaresSimulatorTeam/hotfix/v2.16.8

laurent-laporte-pro committed:
v2.17

Merge pull request #2028 from AntaresSimulatorTeam/release/2.17.0

skamril committed:
v2.17.1

Merge pull request #2057 from AntaresSimulatorTeam/release/v2.17.1

skamril committed:
v2.17.2

Merge pull request #2074 from AntaresSimulatorTeam/release/v2.17.2

skamril committed:
v2.17.3

Merge pull request #2096 from AntaresSimulatorTeam/release/v2.17.3

skamril committed:
v2.17.4

Merge pull request #2104 from AntaresSimulatorTeam/release/v2.17.4

skamril committed:
v2.17.5

Merge pull request #2111 from AntaresSimulatorTeam/release/2.17.5

TheoPascoli committed:
WIP : first commit

TheoPascoli committed:
isort and black

TheoPascoli committed:
add versioning

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add versioning

TheoPascoli committed:
add versioning

TheoPascoli committed:
add versioning

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
correct typing issue

TheoPascoli committed:
fix tests

TheoPascoli committed:
fix tests

TheoPascoli committed:
fix tests

TheoPascoli committed:
refactor code

TheoPascoli committed:
refactor code

TheoPascoli committed:
fix mypy

TheoPascoli committed:
fix mypy

TheoPascoli committed:
fix mypy

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
change ui handling

TheoPascoli committed:
black fixing

TheoPascoli committed:
fix errors

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix errors

TheoPascoli committed:
fix errors

MartinBelthle committed:
refactor code

TheoPascoli committed:
isort and black

TheoPascoli committed:
refactor classes

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

TheoPascoli committed:
add integration tests

skamril committed:
v2.17.6

Merge pull request #2159 from AntaresSimulatorTeam/release/2.17.6

TheoPascoli committed:
add integration tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add integration tests

TheoPascoli committed:
WIP : add update link endpoint

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix: fix tests

skamril committed:
build(ui): fix issue with build result not working (#2163)

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/master' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' of https://github.com/AntaresSimulatorTeam/AntaREST into fix/fix-endpoint-links-creation

TheoPascoli committed:
fix: fix tests

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
fix: remove optional typing in models

TheoPascoli committed:
Merge remote-tracking branch 'origin/fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

TheoPascoli committed:
WIP

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
conflict: merge dev

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
WIP

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
add abstract class for links

TheoPascoli committed:
isort

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
adapt to new handling of study_versions

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

TheoPascoli committed:
refactor series saving and to_dto method

TheoPascoli committed:
refactor series saving and to_dto method

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
add tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -73,11 +104,134 @@ def validate_series(
+        if self.series != other.series:
+            commands.append(
+                ReplaceMatrix(
+                    target=f"@links_series/{area_from}/{area_to}",
```
What is this path ?

TheoPascoli replied:
I'm not sure, it was already here

MartinBelthle replied:
After some research I saw that it's used inside replace matrix (see class AliasDecoder inside utils.py) as an alias to hide the complexity of the versions. Never seen this code before it's pretty unusual but we can keep it. I thought you introduced it

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -73,11 +104,134 @@ def validate_series(
+
+    def get_inner_matrices(self) -> List[str]:
+        list_matrices = []
+        if self.series:
```
You could do a for loop based on `getattr`, see `create_binding_constraints.py` to simplify the code

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -73,11 +104,134 @@ def validate_series(
+            "area2": self.area2,
+            "parameters": self.parameters,
+        }
+        if self.series:
```
Could use `getattr`

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -73,11 +104,134 @@ def validate_series(
+        self.series = self.series or (self.command_context.generator_matrix_constants.get_link(version=version))
+        self.direct = self.direct or (self.command_context.generator_matrix_constants.get_link_direct())
+        self.indirect = self.indirect or (self.command_context.generator_matrix_constants.get_link_indirect())
+        assert type(self.series) is str
```
check with isinstance

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,73 @@
+        area_from, area_to = sorted([self.area1, self.area2])
+
+        current_parameters = study_data.tree.get(["input", "links", area_from, "properties", area_to])
+        current_parameters.update(self.parameters or {})
```
You shouldn't trust `self.parameters` as users can use the command directly and therefore put anything inside.
I'll suggest you do a roundtrtip around Pydantic objects you've introduced to ensure data quality

TheoPascoli replied:
It has already been validated in LinkManagement

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,73 @@
+
+        output, _ = self._apply_config(study_data.config)
+
+        self.save_series(area_from, area_to, study_data, version)
```
Not sure I understood everything but if someone updates a link property without updating a matrix you shouldn't modify them (I think the `save_series` method puts the default value if no value is given and if it does it should change)

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,73 @@
+
+class UpdateLink(AbstractLinkCommand):
+    """
+    Command used to create a link between two areas.
```
Update the doc

MartinBelthle commented on antarest/study/web/study_data_blueprint.py:
```diff
@@ -198,6 +197,24 @@ def create_link(
         return study_service.create_link(uuid, link_creation_info, params)
 
+    @bp.put(
+        "/studies/{uuid}/links",
```
I think we should use `link` and not `links` as it's only used for one link.
On the other hand the creation endpoint you can't change is `links` ...
What do you think @sylvlecl ?

sylvlecl replied:
We should be consistent for all link-related endpoints (can be seen as the "links API"): better to keep `links` here.
However, I think it would make more sense to have either:
- a URL for updating only one link with the link ID in its path:
```studies/{uuid}/links/{link_id}```
- or this URL for updating a list of links

MartinBelthle replied:
Okay. We should probably discuss what's a link id then. Is it `area1%area2` or `area1/area2` or another id ?

MartinBelthle commented on antarest/study/web/study_data_blueprint.py:
```diff
@@ -198,6 +197,24 @@ def create_link(
+        current_user: JWTUser = Depends(auth.get_current_user),
+    ) -> t.Any:
+        logger.info(
+            f"Updating link for study {uuid}",
```
Add info about which link in the logs

MartinBelthle commented on antarest/study/service.py:
```diff
@@ -1856,6 +1855,25 @@ def create_link(
+        assert_permission(params.user, study, StudyPermissionType.WRITE)
+        self._assert_study_unarchived(study)
+        updated_link = self.links.update_link(study, link_update_dto)
+        self.event_bus.push(
```
We usually don't push this inside the eventbus but perhaps it's a good idea. I think it can flood the bus but not sure

TheoPascoli replied:
I saw this on area updating and other services but I can remove it

MartinBelthle replied:
Okay why not

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -51,39 +59,119 @@ class LinkManager:
+    def update_link(self, study: RawStudy, link_creation_info: LinkInfoDTOType) -> LinkInfoDTOType:
+        file_study = self.storage_service.get_storage(study).get_raw(study)
+
+        self.check_attributes_coherence(file_study, int(study.version), link_creation_info)
```
It would be better to use StudyVersion.parse(study.version)

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -137,15 +225,17 @@ def update_links_props(
             cmd = UpdateConfig(
                 target=path,
-                data=properties.to_config(),
+                data=properties.to_ini(int(study.version)),
```
Use StudyVersion.parse(study.version)

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,253 @@
+        preparer = PreparerProxy(client, user_access_token)
+        study_id = preparer.create_study("foo", version=820)
+        if study_type == "variant":
+            study_id = preparer.create_variant(study_id, name="Variant 1")
```
When the study is a variant it could be nice to check the created `update_link` command to ensure it only contains the data you updated and not all the data

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,253 @@
+        }
+        assert expected == res.json()
+
+        # Test update link same area
```
It could be nice to check that the endpoint fails when given wrong parameters

MartinBelthle reviewed (changes requested):
Little review as I saw that you put the ticket in the  `Review` column in JIRA

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
code review changes

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
code review changes

TheoPascoli committed:
code review changes

TheoPascoli committed:
code review changes

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+                exclude_unset=True, by_alias=True
+            )
+        except ValidationError:
+            raise LinkValidationError("One or more fields are forbidden")
```
I'm not sure we should catch this error. Because model_validate will already raise the `ValidationError` which has all info we need and the correct HTTP code I believe. At least you should do this to keep the stacktrace but I prefer not catching:
```python
except ValidationError as e:
     raise LinkValidationError("One or more fields are forbidden") from e
```

TheoPascoli replied:
I removed the catch

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+
+        current_parameters = study_data.tree.get(["input", "links", area_from, "properties", area_to])
+
+        current_parameters.update(properties or {})
```
It's possible to have properties = None after a model_dump ? If not you can remove the `or`

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -0,0 +1,294 @@
+            commands = res.json()
+            command_args = commands[-1]["args"]
+            assert command_args["parameters"]["hurdles-cost"] == False
+            assert "loop-flow" not in command_args["parameters"]
```
You could be more strict and check that the content equals exactly {"parameters": {"hurdles-cost": False}}

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -73,11 +105,128 @@ def validate_series(
+        list_matrices = []
+        for attr in MATRIX_ATTRIBUTES:
+            value = getattr(self, attr, None)
+            if value:
```
You can merge the 2 lines above into only one:
`if value := getattr(self, attr, None)`

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+        output, _ = self._apply_config(study_data.config)
+
+        if any([self.series, self.direct, self.indirect]):
+            self.save_series(area_from, area_to, study_data, version)
```
It's better cause you solved a part of the issue  I explained but I believe there's still an issue.
If someone just passes the series matrix inside the update you shouldn't modify the 2 others.
And I think entering `save_series` that's what you do
You can add a test to ensure this behaviour is correct

MartinBelthle reviewed (changes requested):
Only a few things to say

TheoPascoli committed:
Refactor link creation and updating logic

Split `save_series` method into three distinct methods: `save_series`, `save_direct`, and `save_indirect`. Updated the `update_link` command to handle the new methods and fixed an error message in the integration tests.

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

sylvlecl committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
Refactor to include camelcase to the return object

TheoPascoli committed:
Refactor link DTOs and validation logic

Updated the structure of LinkInfoDTO classes to use Pydantic's ConfigDict for configuration and streamlined validation logic. Removed unnecessary imports and redundant configurations, while ensuring filter fields are checked appropriately for different study versions.

TheoPascoli committed:
Refactor filter values to use a constant list

Use a centralized FILTER_VALUES list for filter_synthesis and filter_year_by_year in LinkInfoProperties820. This change simplifies code maintenance and ensures consistency across filter validations.

TheoPascoli committed:
Refactor filter fields to use enumerations

Replaced filter fields with a comma-separated list of enums for better validation and clarity. This enhances type safety and reduces the risk of invalid filter values being used. The previous string-based validation approach has been removed in favor of utility functions for converting and serializing enum lists.

TheoPascoli committed:
Add a missing newline before LinkInfoProperties820 class

Ensure code readability and consistency by adding a newline before the definition of the `LinkInfoProperties820` class. This change improves the visual separation and organization of the code.

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
Add a missing newline before LinkInfoProperties820 class

Ensure code readability and consistency by adding a newline before the definition of the `LinkInfoProperties820` class. This change improves the visual separation and organization of the code.

TheoPascoli committed:
Remove unused import and update type annotations

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
Refactor method names in update_link module

TheoPascoli committed:
Rename variables to reflect update context in link management

TheoPascoli committed:
Refactor `get_one_link` to `get_updated_link`

TheoPascoli committed:
Refactor link update logic in link_management.

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/business/table_mode_management.py

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation

TheoPascoli committed:
feat: refactor link management to use new LinkDTO class

TheoPascoli committed:
feat: simplify JSON dumping in create_link

TheoPascoli committed:
Refactor link management imports and optimize validations

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation-refacto' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
feat: refactor link update process for improved consistency

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
feat: refactor link path handling in link management.

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
feat: remove unused LINK_PATH and inline its usage

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

skamril committed:
refactor(ui-api): update links endpoints with new DTO format

skamril committed:
fix(ui-api): update links endpoints with new DTO format

skamril committed:
refactor(ui-api): update links endpoints with new DTO format

TheoPascoli committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

TheoPascoli committed:
feat: add header

TheoPascoli committed:
feat: add header

TheoPascoli committed:
feat: update return type of get_links function

skamril committed:
Merge branch 'dev' into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
feat: update return type of get_links function

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation-refacto' of https://github.com/AntaresSimulatorTeam/AntaREST into fix/fix-endpoint-links-creation-refacto

TheoPascoli committed:
fix: fix color field validation in link model

TheoPascoli committed:
fix: correct color value validation

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation-refacto' into feature/add-endpoint-link-update

TheoPascoli committed:
refactor(link): update DTO format

TheoPascoli committed:
refactor(link): update DTO format

TheoPascoli committed:
fix: fix test assertion message for input length validation

TheoPascoli committed:
fix: fix test assertion message for input length validation

TheoPascoli committed:
fix: add copyright and license headers

TheoPascoli committed:
fix: add copyright and license headers

TheoPascoli committed:
fix: add copyright and license headers

TheoPascoli committed:
fix: add copyright and license headers

TheoPascoli committed:
fix: adjust color field validation limits

TheoPascoli committed:
fix: adjust color field validation limits

TheoPascoli committed:
Merge branch 'dev' into feat/ANT-2312_ANT-1882_links-endpoints

TheoPascoli committed:
fix: refactor and simplify link validation and creation following github review

TheoPascoli committed:
fix: refactor and simplify link validation and creation following github review

TheoPascoli committed:
Merge branch 'feat/ANT-2312_ANT-1882_links-endpoints' of https://github.com/AntaresSimulatorTeam/AntaREST into feat/ANT-2312_ANT-1882_links-endpoints

TheoPascoli committed:
feat: refactor type annotations and version parsing

TheoPascoli committed:
feat: refactor condition handling in table_mode_management

TheoPascoli committed:
Merge branch 'fix/fix-endpoint-links-creation-refacto' into feature/add-endpoint-link-update

TheoPascoli committed:
fix: missing newline at end of file in model init

TheoPascoli committed:
Merge branch 'feat/ANT-2312_ANT-1882_links-endpoints' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/business/model/link_model.py
#	antarest/study/business/table_mode_management.py
#	antarest/study/storage/rawstudy/model/filesystem/config/links.py
#	antarest/study/storage/variantstudy/model/command/create_link.py
#	tests/integration/study_data_blueprint/test_link.py
#	tests/storage/business/test_arealink_manager.py

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
fix: refactor link model and update related tests

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/business/link_management.py
#	antarest/study/business/model/link_model.py
#	antarest/study/business/table_mode_management.py
#	antarest/study/storage/variantstudy/model/command/create_link.py
#	tests/integration/study_data_blueprint/test_link.py

TheoPascoli committed:
refactor: remove duplicate and unused imports

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

# Conflicts:
#	antarest/study/storage/variantstudy/model/command/create_link.py

TheoPascoli committed:
feat: add study_version to CommandDTO in create_link

TheoPascoli committed:
Merge branch 'dev' into feature/add-endpoint-link-update

MartinBelthle commented on antarest/study/web/study_data_blueprint.py:
```diff
@@ -197,6 +197,25 @@ def create_link(
+    )
+    def update_link(
+        uuid: str,
+        link_creation_info: LinkDTO,
```
This keyword should change as we are in an update endpoint

MartinBelthle commented on antarest/study/business/model/link_model.py:
```diff
@@ -98,5 +98,20 @@ class LinkInternal(AntaresBaseModel):
     def to_dto(self) -> LinkDTO:
-        data = self.model_dump()
-        return LinkDTO(**data)
+        return LinkDTO(
```
We should keep the 2 lines with model_dump

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+        try:
+            file_study.tree.get(["input", "links", area_from, "properties", area_to])
+        except KeyError:
+            raise LinkValidationError(f"The link {area_from} -> {area_to} is not present in the study")
```
There's an exception LinkNotFoundError defined inside `xpansion_management.py` that we should move into `exceptions.py` and use it here I believe

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+    def get_ini_link(self, study: RawStudy, link: LinkInternal) -> LinkInternal:
+        file_study = self.storage_service.get_storage(study).get_raw(study)
+
+        area_from, area_to = sorted([link.area1, link.area2])
```
Shouldn't we do this sorting when creating the LinkDTO object (with pydantic model_validator) to avoid doing this every time ?

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+            area1=link.area1,
+            area2=link.area2,
+            parameters=link.model_dump(
+                include=link_dto.model_fields_set, exclude={"area1", "area2"}, exclude_none=True
```
Why did you use the `include` keyword here ?

TheoPascoli replied:
We must include only the fields that has been specified by the client, otherwise we would replace the existing values by the default ones

MartinBelthle replied:
Okay I see I didn't know model_fields_set worked like this, it's nice

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+        link_properties.update({"area1": area_from, "area2": area_to})
+        updated_link = LinkInternal.model_validate(link_properties)
+
+        return link.model_copy(update=updated_link.model_dump())
```
Why don't you simply return the updated_link you got line 120 ?

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+        except KeyError:
+            raise LinkValidationError(f"The link {area_from} -> {area_to} is not present in the study")
+
+    def get_ini_link(self, study: RawStudy, link: LinkInternal) -> LinkInternal:
```
Opinion and really minor but: It could be good to rename this method `get_internal_link` to hide the fact that we get the data inside a ini file as it could change

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+                status=True,
+                message=f"Link between '{self.area1}' and '{self.area2}' updated",
+            ),
+            {"area_from": area_from, "area_to": area_to},
```
Just a question but where do we use the 2nd part of the OutputTuple ?

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+
+        properties = study_data.tree.get(["input", "links", area_from, "properties", area_to])
+
+        new_properties = LinkInternal.model_validate(self.parameters).model_dump(include=self.parameters, by_alias=True)
```
Same question about include here.

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,83 @@
+        return super().to_dto()
+
+    def match_signature(self) -> str:
+        return str(
```
You could put that inside the AbstractCommand and use it inside the create too

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +53,128 @@ def validate_series(
+        return CommandDTO(action=self.command_name.value, args=args, study_version=self.study_version)
+
+    def match(self, other: ICommand, equal: bool = False) -> bool:
+        if not isinstance(other, CreateLink):
```
Should probably be replaced with `if not isinstance(other, self.__class__):` as you're inside AbstractCommand

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +53,128 @@ def validate_series(
+        )
+
+    def _create_diff(self, other: "ICommand") -> List["ICommand"]:
+        other = cast(CreateLink, other)
```
Should be casted inside Abstract I believe to work with UpdateCommand

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +53,128 @@ def validate_series(
+        return list_matrices
+
+    def save_series(self, area_from: str, area_to: str, study_data: FileStudy, version: StudyVersion) -> None:
+        self.series = self.series or (self.command_context.generator_matrix_constants.get_link(version=version))
```
You don't have to do this anymore as when you enter this function you're sure you have series.
Same goes for save_direct and save_indirect

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +53,128 @@ def validate_series(
+                self.series,
+                ["input", "links", area_from, f"{area_to}_parameters"],
+            )
+            study_data.tree.save({}, ["input", "links", area_from, "capacities"])
```
Not really sure why you do this, is it to ensure we don't have both matrices writing ? I think you should remove this. 
Same goes for the 2 save_ functions

TheoPascoli replied:
I'm not sure why. It was there before. I removed it

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -19,6 +20,117 @@
+                ],
+            )
+            assert res.status_code == 500
+            expected = {
```
You shouldn't check the entire content as if pydantic changes its error msg we'll have to change this. You could just check our msg: "Unexpected exception occurred when trying to apply command CommandName.UPDATE_LINK: 1 validation error for LinkInternal"

MartinBelthle commented on tests/integration/study_data_blueprint/test_link.py:
```diff
@@ -19,6 +20,117 @@
+            commands = res.json()
+            command_args = commands[-1]["args"]
+            assert command_args["parameters"] == {"hurdles_cost": False}
+
```
it could be nice to add tests for matrices as I don't see any.

TheoPascoli committed:
fix: refactor link management and update error handling

TheoPascoli committed:
Merge remote-tracking branch 'origin/dev' into feature/add-endpoint-link-update

TheoPascoli committed:
fix: refactor to_dto method and fix test string format

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+        try:
+            link_properties = file_study.tree.get(["input", "links", area_from, "properties", area_to])
+        except KeyError:
+            raise LinkValidationError(f"The link {area_from} -> {area_to} is not present in the study")
```
You should also raise a LinkNotFound here.

Also, you could perhaps call check_link_existence inside this method to avoid duplicating code (needs a little refactoring to return the properties inside the check method).

MartinBelthle commented on antarest/study/business/link_management.py:
```diff
@@ -76,6 +78,49 @@ def create_link(self, study: Study, link_creation_dto: LinkDTO) -> LinkDTO:
+        return updated_link.to_dto()
+
+    def check_link_existence(self, file_study: FileStudy, link: LinkInternal) -> None:
+        area_from, area_to = sorted([link.area1, link.area2])
```
You shouldn't have to sort this anymore as you did it inside the Area pydantic class

Same goes for other parts of the code I'll let you check

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/update_link.py:
```diff
@@ -0,0 +1,78 @@
+from antarest.study.storage.rawstudy.model.filesystem.factory import FileStudy
+from antarest.study.storage.variantstudy.model.command.common import CommandName, CommandOutput
+from antarest.study.storage.variantstudy.model.command.create_link import AbstractLinkCommand
+from antarest.study.storage.variantstudy.model.command.icommand import MATCH_SIGNATURE_SEPARATOR, ICommand, OutputTuple
```
Remove unused import MATCH_SIGNATURE_SEPARATOR

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +53,130 @@ def validate_series(
+    def save_direct(self, area_from: str, area_to: str, study_data: FileStudy, version: StudyVersion) -> None:
+        self.direct = self.command_context.generator_matrix_constants.get_link_direct()
+        assert isinstance(self.direct, str)
+        if version >= STUDY_VERSION_8_2:
```
Not really the point of this dev but we should raise an Exception when someone gives "direct" or "indirect" with a study prior to v8.2
This could be validated inside a field_validator now that we have the attribute `study_version` inside the commands.
This way you could remove the version check inside the 2 functions
It could also be good to add tests for this inside your matrix tests

MartinBelthle replied:
In fact it could break the compatibility of the R scripts. We should ask them how they call this command before

TheoPascoli replied:
After checking with script team, I removed those conditions and added a model validator

MartinBelthle reviewed (changes requested):
Really minor changes, otherwise looks good to me. With the matrix tests we'll be good to go

TheoPascoli committed:
fix: changes after review

TheoPascoli committed:
fix: changes after review

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +54,138 @@ def validate_series(
         if self.area1 == self.area2:
             raise ValueError("Cannot create link on same node")
+
+        if StudyVersion.parse(self.study_version) < STUDY_VERSION_8_2 and (
```
Don't need to parse it's already a StudyVersion object

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +54,138 @@ def validate_series(
+        if StudyVersion.parse(self.study_version) < STUDY_VERSION_8_2 and (
+            self.series is not None or self.direct is not None or self.indirect is not None
+        ):
+            raise LinkValidationError(
```
I believe you shouldn't raise an error when series are given as in the code we do handle this case. So I think you should only check the 2 other matrices and put back the tests as they were

MartinBelthle commented on antarest/study/storage/variantstudy/model/command/create_link.py:
```diff
@@ -57,11 +54,138 @@ def validate_series(
+        return list_matrices
+
+    def save_series(self, area_from: str, area_to: str, study_data: FileStudy, version: StudyVersion) -> None:
+        self.series = self.command_context.generator_matrix_constants.get_link(version=version)
```
The 3 save method have the same issue. You replace `self.series` by the default matrix but if someone gives you a matrix you should use the one that was given. You should remove this line

MartinBelthle reviewed (changes requested):
Still got little issues

TheoPascoli committed:
feat: fix tests

TheoPascoli committed:
feat: change the url for the update endpoint.
Added new linkDtoForUpdate object to hide area1 and area2 from the client

MartinBelthle commented on antarest/study/business/model/link_model.py:
```diff
@@ -78,6 +86,11 @@ def to_internal(self, version: StudyVersion) -> "LinkInternal":
 
 
+class LinkDtoForUpdate(LinkDTO):
+    area1: SkipJsonSchema[str] = Field("a", exclude=True)
```
Weird way to do it. Why did you introduce fields `a` and `b` ?

MartinBelthle replied:
After testing this allows user to specfiy area1 and area2 inside the body and they we'll be ignored. I don't think we want that. I'd prefer you introduce a class that just has the base attributes and make LinkDTO inherit from this :/

TheoPascoli replied:
I changed it

MartinBelthle commented on antarest/study/business/model/link_model.py:
```diff
@@ -44,11 +45,18 @@ class Area(AntaresBaseModel):
 
 class LinkDTO(Area):
-    model_config = ConfigDict(alias_generator=to_camel_case, populate_by_name=True, extra="forbid")
+    model_config = ConfigDict(
```
Could be in one line

TheoPascoli replied:
Indeed

MartinBelthle commented on antarest/study/business/model/link_model.py:
```diff
@@ -78,6 +86,11 @@ def to_internal(self, version: StudyVersion) -> "LinkInternal":
         return LinkInternal(**data)
 
 
+class LinkDtoForUpdate(LinkDTO):
```
Could be DTO instead of Dto to fit with existing class

MartinBelthle reviewed (commented):
2 minor comments that don't really need to be adress and 1 question

TheoPascoli committed:
feat: change model logic for update endpoint

TheoPascoli committed:
feat: change model logic for update endpoint
